File: src/app/commands/set_shortcut.ts
import { register, unregister } from '@tauri-apps/plugin-global-shortcut';
import { getStore } from '@tauri-apps/plugin-store';
import onShortcut from '../handlers/on_shortcut';
import type { AppConfig } from '../state';
import { APP_STORE_FILE } from '../constants';

export async function setShortcut(keybind: string): Promise<boolean> {
	// get store
	const store = await getStore(APP_STORE_FILE);
	if (!store) {
		throw new Error('Failed to get store');
	}

	// get config
	let config = await store.get<AppConfig>('config');
	if (!config) {
		throw new Error('Failed to get App Config');
	}

	console.log(Object.entries(config));

	// get old keybind, set new keybind
	const old_keybind = config.shortcut;
	try {
		// unregister old shortcut
		if (old_keybind) {
			await unregister(old_keybind);
		}

		// attempt to register new shortcut
		await register(keybind, onShortcut);
		console.log('Set chatbar shortcut to', keybind);
		return true;
	} catch {
		console.warn(`Shortcut ${keybind} is not valid, using ${old_keybind}`);
		// re-register old shortcut
		await register(old_keybind, onShortcut);
		return false;
	}
}

File: src-tauri/src/commands/set_shortcut.rs
use crate::{handlers::on_shortcut, prelude::*};
use serde_json::json;
use tauri_plugin_global_shortcut::{GlobalShortcutExt, Shortcut, ShortcutState};
use tauri_plugin_store::StoreExt;

use crate::state::AppConfig;

pub fn set_shortcut_internal(app_handle: &tauri::AppHandle, new_shortcut: String) -> Result<()> {
    // Try to convert the string to a shortcut to validate it
    let shortcut: Shortcut = new_shortcut
        .clone()
        .try_into()
        .map_err(|_| "Invalid shortcut format".to_string())?;

    // Update store
    let store = app_handle
        .store("config.json")
        .map_err(|_| "Failed to get store")?;

    // Get config from store
    let config = store
        .get("config")
        .unwrap_or_else(|| json!(AppConfig::default()));
    let mut config: AppConfig =
        serde_json::from_value(config).map_err(|_| "Failed to parse config".to_string())?;

    config.shortcut = new_shortcut;

    store.set("config", json!(config));

    store.save().map_err(|e| e.to_string())?;

    // Unregister all shortcuts and register the new one
    let global_shortcut = app_handle.global_shortcut();
    global_shortcut
        .unregister_all()
        .map_err(|e| e.to_string())?;

    let handle = app_handle.clone();
    global_shortcut
        .on_shortcut(shortcut, move |_, _, event| {
            if event.state == ShortcutState::Pressed {
                on_shortcut(&handle).unwrap();
            }
        })
        .map_err(|e| e.to_string())?;

    Ok(())
}

#[tauri::command]
pub async fn set_shortcut(
    app_handle: tauri::AppHandle,
    new_keybind: String,
) -> std::result::Result<(), String> {
    set_shortcut_internal(&app_handle, new_keybind).map_err(|e| e.to_string())
}
